Automated Market Maker (XLS-30d) reference documentation (#1541)
XRPLF/xrpl-dev-portal: content/references/protocol-reference/transactions/transaction-types/ammbid.md

- If the auction slot is currently empty, expired, or in its last interval, the **minimum bid** is **0.001% of the AMM's total LP Tokens balance**.

    **Caution:** This minimum value is a placeholder and may change before the AMM feature becomes finalized.

- Otherwise, the price to outbid the current holder is calculated using the following formula:

        P = B × 1.05 × (1 - t⁶⁰) + M

    - `P` is the price to outbid, in LP Tokens.
    - `B` is the price of the current bid, in LP Tokens.
    - `t` is the fraction of time elapsed in the current 24-hour slot, rounded down to a multiple of 0.05.
    - `M` is the **minimum bid** as defined above.

    There are two special cases for the cost to outbid someone. In the **first interval** after someone wins the bid, the price to outbid them is only 5% more than their bid:

>         P = B × 1.05

Reviewer: It's updated to `P = B*1.05 + M`.

---
html: ammbid.html
parent: transaction-types.html
blurb: Bid on an Automated Market Maker's auction slot, which grants a discounted fee.
labels:
  - AMM
status: not_enabled
---
# AMMBid
[[Source]](https://github.com/gregtatcam/rippled/blob/amm-core-functionality/src/ripple/app/tx/impl/AMMBid.cpp "Source")
<!-- TODO: Update source link to merged version when available -->

{% include '_snippets/amm-disclaimer.md' %}

> Bid on an Automated Market Maker's (AMM's) auction slot. If this transaction succeeds, you win the bid and can trade against the AMM at a discounted fee until you are outbid or 24 hours have passed. If you are outbid before 24 hours have passed, you are refunded part of the cost of your bid based on how much time remains.

Reviewer 2: ```suggestion
Bid on an Automated Market Maker's (AMM's) auction slot. If you win, you can trade against the AMM at a discounted fee until you are outbid or 24 hours have passed. If you are outbid before 24 hours have passed, you are refunded part of your bid based on how much time remains.
```
Can we remove explicitly mentioning this is a transaction since we're already under transaction types in page hierarchy.

Author: Hm, I'm not sure how I feel about this suggestion. I like that it's more succinct. My thought behind the original phrasing is that I want to answer the questions, "how do you know if you won?" or "what does it mean for this transaction to be successful?" since there are cases where "success" at the level of executing the transaction doesn't necessarily mean what you might think. (For example, when placing an Offer in the DEX, "success" might mean, "your Offer has been recorded and is waiting for a match", not "you successfully traded currency". Or when canceling a prior Offer, "success" means "there is no longer any such Offer" which can mean "it has been canceled" or "it was already fully consumed" as well as "it never existed because you made a typo".) So I do think it's useful to convey, "if the transaction has a success result code, that _actually_ means you won the bid" and not just, "your bid was received, but it may or may not have been too low to win.")

Reviewer 2: Ah okay. The distinction on what a successful transaction means makes sense to include.

| `BidMax`       | [Currency Amount][] | Amount            | No        | Pay at most this amount for the slot. If the cost to win the bid is higher than this amount, the transaction fails. If omitted, pay as much as necessary to win the bid. |
| `AuthAccounts` | Array               | STArray           | No        | A list of up to 4 additional accounts that you allow to trade at the discounted fee. This cannot include the address of the transaction sender. Each of these objects should be an [AuthAccount object](#authaccount-objects). |

You cannot specify both `BidMin` and `BidMax`.

### AuthAccount Objects

Each member of the `AuthAccounts` array must be an object with the following field:

| Field          | JSON Type | [Internal Type][] | Required? | Description |
|:---------------|:----------|:------------------|:----------|:------------|
| `Account`      | String    | AccountID         | Yes       | The address of the account to authorize. |

Like other "inner objects" that can appear in arrays, the JSON representation of each of these objects is wrapped in an object whose only key is the object type, `AuthAccount`.

> ## Slot Price Calculations

Reviewer 2: Same suggestion as the one on refunds. Maybe rename this to Auction Slot Prices?

> If successful, the transaction automatically outbids the previous slot owner and debits the bid price from the sender's LP Tokens. The price to win the auction decreases over time, divided into 20 intervals of 72 minutes each. If the sender does not have enough LP Tokens to win the bid, or the price of the bid is higher than the transaction's `BidMax` value, the transaction fails with a `tecAMM_FAILED_BID` result.

Reviewer 2: I might be missing something, but it doesn't seem like the 20 intervals detail comes into play except in the first and last interval. As written I assumed the current price decrease happens at each 72 minute interval, but the formula used just takes a fraction of 24 hour periods.

I've made several suggestions to maybe simplify this and just say the first/last 72 minutes. Feel free to disregard if the interval thing is important.

Author: My understanding is that the formula uses a _rounded fraction_ of 24 hour periods. The rounding means that the price increase _does_ happen at each interval.

Also, the `amm_info` method reports what interval the current winning bid is in, so the intervals concept is useful to understand for that purpose as well, I think.

I'm going to think more on this to decide which way to go on the intervals thing.

Reviewer 2: I wasn't aware of the amm_info method reporting in intervals, in which case I would leave as is.

> - If the auction slot is currently empty, expired, or in its last interval, the **minimum bid** is **0.001% of the AMM's total LP Tokens balance**.

Reviewer 2: ```suggestion
- If the auction slot is empty, expired, or in its last 72 minutes, the **minimum bid** is **0.001% of the AMM's total LP Tokens balance**.
```

Author: Decided after thinking about it more not to change the intervals thing.

    **Caution:** This minimum value is a placeholder and may change before the AMM feature becomes finalized.

- Otherwise, the price to outbid the current holder is calculated using the following formula:

>         P = B × 1.05 × (1 - t⁶⁰) + M

Reviewer 2: Am I understanding this right? t⁶⁰ is elapsed time to the 60th power?

Author: line 91: elapsed fraction of time, rounded to a multiple of 0.05, and then raised to the 60th power. So for example in the third interval, t=0.15 therefore t⁶⁰ = `3.67e-50` (where `e-50` means "times 10 raised to the negative-50th power") Something like that.

    - `P` is the price to outbid, in LP Tokens.
    - `B` is the price of the current bid, in LP Tokens.
    - `t` is the fraction of time elapsed in the current 24-hour slot, rounded down to a multiple of 0.05.
    - `M` is the **minimum bid** as defined above.

>     There are two special cases for the cost to outbid someone. In the **first interval** after someone wins the bid, the price to outbid them is the minimum bid plus 5% more than the existing bid:

Reviewer 2: ```suggestion
There are two special cases for the cost to outbid someone. In the first 72 minutes after someone wins the bid, the price to outbid them is the minimum bid plus 5% more than the existing bid:
```

        P = B × 1.05 + M

>     In the **last interval** of someone's slot, the cost to outbid someone is only the minimum bid:

Reviewer 2: ```suggestion
In the last 72 minutes of someone's slot, the cost to outbid someone is only the minimum bid:
```

        P = M

**Note:** To make sure all servers in the network reach the same results when building a ledger, time measurements are based on the [official close time](ledgers.html#ledger-close-times) of the previous ledger, which is approximate.

> ## Refunds

Reviewer 2: Maybe rename  this header to specify to Bid Refunds or Auction Slot Refunds?

> In the case where the transaction outbids the previous slot holder, the AMM refunds the previous holder part of the price of the auction slot. The amount refunded is calculated using the following formula:

Reviewer 2: ```suggestion
When a transaction outbids an auction slot holder, the AMM refunds the previous holder using this formula:
```
Just a suggestion for brevity.

```text
R = B × (1 - t)
```

- `R` is the amount to refund, in LP Tokens.
- `B` is the price of the previous bid to be refunded, in LP Tokens.
- `t` is the fraction of time elapsed in the current 24-hour slot, rounded down to a multiple of 0.05.

> As a special case, during the final (20th) interval of the auction slot, the refunded amount is zero.

Reviewer 2: ```suggestion
**Note:** No refund is issued during the last 72 minutes of an auction slot.
```
